domain/service/employees.go

```diff
@@ -0,0 +1,187 @@
+package service
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"io"
+	"time"
+
+	"github.com/Shelex/grpc-go-demo/domain/entities"
+	"github.com/Shelex/grpc-go-demo/proto"
+	"github.com/Shelex/grpc-go-demo/storage"
+	"github.com/Shelex/grpc-go-demo/storage/documents"
+	"github.com/google/uuid"
+	log "github.com/sirupsen/logrus"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/metadata"
+)
+
+type employeeService struct {
+	repository storage.Storage
+	documents  documents.FileStorage
+	grpc       *grpc.Server
+}
+
+func (e *employeeService) Employees(req *proto.GetAllRequest, stream proto.EmployeeService_EmployeesServer) error {
+	employees, err := e.repository.GetAll()
+	if err != nil {
+		return fmt.Errorf("failed to get employees: %w", err)
+	}
+	for _, emp := range employees {
+		if err := stream.Send(&proto.EmployeeResponse{Employee: entities.EmployeeFromStorageToProto(emp)}); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (e *employeeService) EmployeeByID(ctx context.Context, req *proto.ByIDRequest) (*proto.EmployeeResponse, error) {
+	if md, ok := metadata.FromIncomingContext(ctx); ok {
+		log.Printf("metadata received: %v\n", md)
+	}
+	employee, err := e.repository.GetEmployee(req.ID)
+	if err != nil {
+		return nil, err
+	}
+	return &proto.EmployeeResponse{
+		Employee: entities.EmployeeFromStorageToProto(employee),
+	}, nil
+}
+
+func (e *employeeService) AddEmployee(ctx context.Context, req *proto.EmployeeRequest) (*proto.EmployeeResponse, error) {
+	req.Employee.ID = uuid.New().String()
+	employee, err := e.repository.AddEmployee(entities.EmployeeFromProtoToStorage(req.Employee))
+	if err != nil {
+		return nil, fmt.Errorf("failed to save employee: %w", err)
+	}
+	log.Printf("employee %s successfully saved;\n", employee.ID)
+	return &proto.EmployeeResponse{Employee: entities.EmployeeFromStorageToProto(employee)}, nil
+}
+func (e *employeeService) AddEmployees(stream proto.EmployeeService_AddEmployeesServer) error {
+	var savedCount int
+	var errorMessage string
+	for {
+		emp, err := stream.Recv()
+		if err == io.EOF {
+			break
+		}
+		if err != nil {
+			return err
+		}
+		emp.Employee.ID = uuid.New().String()
+		saved, err := e.repository.AddEmployee(entities.EmployeeFromProtoToStorage(emp.Employee))
+		if err != nil {
+			errorMessage += fmt.Sprintf("\n%s", err.Error())
```

Comment:
Ive recently experienced this issue too... do you collect up all your errors and send them at stream close, or do you fail on the first error, or do you send back results in a bi-directional stream function.

Some streams are long-lived, so it might be the case that holding back all errors until the end, might hide errors from the client until it is too late. In this case, Im guessing that the steram is fairly short-lived, so it might be ok to keep it like this. However, I would say you need some way of identifying which error relates to which entity. One approach is to use an index number.

Reply:
Basically, here it is used in order to validate bulk addition of users which would be validated. For now it is only field "badgeNumber".  So in case of error it just gathers failed additions and returns such structure:
```
type EmployeeSaveResult struct {
	SavedEmployees []*Employee `json:"savedEmployees"`
	Errors         *string     `json:"errors"`
}
```

Reply:
@the4thamigo-uk
Have also found nice article on this purpose: [go-multiple-errors-management](https://medium.com/a-journey-with-go/go-multiple-errors-management-a67477628cf1)